sdk/core/azure-core/src/main/java/com/azure/core/implementation/ResponseConstructorsCache.java

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.implementation;
+
+import com.azure.core.http.HttpHeaders;
+import com.azure.core.http.HttpRequest;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.rest.Response;
+import com.azure.core.implementation.serializer.HttpResponseDecoder;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.Exceptions;
+import reactor.core.publisher.Mono;
+
+import java.lang.invoke.LambdaMetafactory;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Constructor;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * A concurrent cache of {@link Response} constructors.
+ */
+final class ResponseConstructorsCache {
+    private final ClientLogger logger = new ClientLogger(ResponseConstructorsCache.class);
+    private final Map<Class<?>, Optional<ResponseConstructor>> cache = new ConcurrentHashMap<>();
```

Comment:
We shouldn't need to have the `ResponseConstructor` wrapped with an `Optional` here. If it is in the map, we should be certain it exists.

Reply:
agree, given this is in hot path, we needs to be careful about any extra allocation. I'm seeing that  even usage of Stream and associated operators for doing this very basic logic (filter, sorting an array with few element [in worst case 3 elements] and mappings) is causing allocation and adding overhead. 

Interestingly, there is already a bug in the current code base, we do sorting in ascending order and take the first, in reality we need to take the last because out of ctrs with 3 to 5 params we need the one with most number of params.